File: leetcode/max-points-on-a-line.py
import math
from typing import List
from collections import Counter

class Solution:
    def get_ab_raw(p0, p1):
        # y = ax + b
        # a = (y1 - y0) / (x1 - x0)
        # b = y - ax
        if p0[0] > p1[0]:
            p = p0
            p0 = p1
            p1 = p
        if p1[0] != p0[0]:
            dy, dx = p1[1] - p0[1], p1[0] - p0[0]
            a = dy / dx
            b = p0[1] - a * p0[0]
        else:
            a = float("inf")
            b = p0[0]
        print(p0, p1, " => ", a, b)
        return a, b
    def get_ab(p0, p1):
        # y = ax + b
        # a = (y1 - y0) / (x1 - x0)
        # b = y - ax
        if p0[0] > p1[0]:
            p = p0
            p0 = p1
            p1 = p
        if p1[0] != p0[0]:
            dy, dx = p1[1] - p0[1], p1[0] - p0[0]
            gcd = math.gcd(dx, dy)
            a = dx / gcd, dy / gcd
            divisor = dx
            dividende = dx * p0[1] - dy * p0[0]
            gcd = math.gcd(divisor, dividende)
            b = divisor / gcd, dividende / gcd
        else:
            a = float("inf")
            b = p0[0]
        print(p0, p1, " => ", a, b)
        return a, b
    def maxPoints0(self, points: List[List[int]]) -> int:
        if len(points) < 3:
            return len(points)
        ret = 2
        for idx, p0 in enumerate(points[:-1]):
            counter = Counter()
            print(p0)
            for p1 in points[idx+1:]:
                a, b = Solution.get_ab(p0, p1)
                # print("p0 %s - p1 %s => %f, %f" % (p0, p1, a, b))
                # if a != float("inf"):
                # counter[(int(a * 10000), int(b*10000))] += 1
                # else:
                counter[(a, b)] += 1
                # p0 = points[idx]
            ret = max(counter.most_common(1)[0][1] + 1, ret)
            print(ret)
            return ret
            #print(counter)
        #print(ret)
    def get_slope(dx, dy):
        if dx < 0:
            dx = -dx
            dy = -dy
        elif dx == 0:
            dy = dy if dy > 0 else -dy  
        gcd = math.gcd(dx, dy)
        return dy / gcd, dx / gcd

    def maxPoints(self, points: List[List[int]]) -> int:
        if len(points) < 3:
            return len(points)
        ret = 2
        for idx, p0 in enumerate(points[:-1]):
            if ret >= len(points) - idx:
                break
            counter = Counter()
            for p1 in points[idx+1:]:
                slope = Solution.get_slope(p0[0] - p1[0], p0[1] - p1[1])
                # print(p0, p1, " => ", slope)
                counter[slope] += 1
            # print(counter.most_common(1))
            ret = max(counter.most_common(1)[0][1] + 1, ret)
        return ret


def check_solution():
    s = Solution()
    points = [[1,1],[2,2],[3,3]]
    assert(s.maxPoints(points) == 3)

    points = [[1,1],[3,2],[5,3],[4,1],[2,3],[1,4]]
    assert(s.maxPoints(points) == 4)

    points = [[1,1],[3,2],[5,1],[4,1],[2,1],[-1,1]]
    assert(s.maxPoints(points) == 5)

    points = [[6,1],[6,2],[6,3],[6,4],[6,0],[-1,1]]
    assert(s.maxPoints(points) == 5)

    points = [[3,3],[1,4],[1,1],[2,1],[2,2]]
    assert(s.maxPoints(points) == 3)

    points = [[54,153],[1,3],[0,-72],[-3,3],[12,-22],[-60,-322],[0,-5],[-5,1],[5,5],[36,78],[3,-4],[5,0],[0,4],[-30,-197],[-5,0],[60,178],[0,0],[30,53],[24,28],[4,5],[2,-2],[-18,-147],[-24,-172],[-36,-222],[-42,-247],[2,3],[-12,-122],[-54,-297],[6,-47],[-5,3],[-48,-272],[-4,-2],[3,-2],[0,2],[48,128],[4,3],[2,4]]
    # import matplotlib.pyplot as plt
    # import numpy as np
    # xaxis = np.array([x for x, y in points[:10]])
    # yaxis = np.array([y for x, y in points[:10]])
    # plt.scatter(xaxis, yaxis)
    # plt.show()
    assert(s.maxPoints(points) == 18)

check_solution()

File: leetcode/max-points-on-a-line.rs
extern crate json;

use std::collections::HashMap;

use algo_tools::load_json_tests;

struct Solution;

impl Solution {
    fn gcd(a: i32, b: i32) -> i32 {
        match b {
            0 => a,
            _ => Solution::gcd(b, a % b)
        }
    }
    fn slope(mut dx : i32, mut dy : i32) -> (i32, i32) {
        if dx < 0 {
            dx = -dx;
            dy = -dy;
        } else if dx == 0 {
            if dy < 0 {
                dy = -dy;
            }
        }
        let gcd = Solution::gcd(dx, dy);
        (dy / gcd, dx / gcd)
    }
    pub fn max_points(points: Vec<Vec<i32>>) -> i32 {
        let l : usize = points.len();
        if l < 3 {
            return l as i32;
        }
        let mut ret = 2;
        for (idx, p0) in points[0..l-1].iter().enumerate() {
            if ret >= l - idx {
                break
            }
            let mut frequency: HashMap<(i32, i32), u32> = HashMap::new();
            for p1 in points[idx+1..l].iter() {
                let slope = Solution::slope(p0[0] - p1[0], p0[1] - p1[1]);
                *frequency.entry(slope).or_insert(1) += 1;
                ret = std::cmp::max(ret, frequency[&slope] as usize);
                // println!("p0{:?}, p1{:?} => slope {:?}", p0, p1, slope);
            }
            // println!("freq: {:?}", frequency);
        }
        ret as i32
    }
}

fn run_test_case(test_case: &json::JsonValue) -> i32 {

	let _i = &test_case["in"];
	let expected = test_case["expected"].as_i32().unwrap();
    let mut i = Vec::new();

	for jpoint in _i.members() {
        let mut point = Vec::new();
        for coord in jpoint.members() {
		    point.push(coord.as_i32().unwrap());
        }
        i.push(point);
	}

	let result = Solution::max_points(i.clone());

	if result == expected {
		return 0;
	}

	println!("max_points({:?}) returned {:?} but expected {:?}\n",
		i, result, expected);
	1
}

fn main() {
	let (tests, test_idx) = load_json_tests();
	let (mut successes, mut failures) = (0, 0);
	if test_idx >= tests.len() as i32 {
		println!("Wrong index {}, only {} tests available!!", test_idx, tests.len());
		return
	}
	if test_idx != -1 {
		let rc = run_test_case(&tests[test_idx as usize]);
		if rc == 0 { successes += 1; }
		else { failures += 1; }
	} else {
		println!("{} tests specified", tests.len());
		for i in 0..tests.len() {
			let rc = run_test_case(&tests[i]);
			if rc == 0 { successes += 1; }
			else { failures += 1; }
		}
	}

	if failures > 0 {
		println!("{} tests succeeded and {} tests failed!!", successes, failures);
	} else {
		println!("All {} tests succeeded!!", successes);
	}
}